### PR TITLE
free relay on webrtc upgrade

### LINKED ISSUE #208: Add relay slots / basic QoS
## Context

Hopr nodes that are willing to become a relay for other can announce themself on-chain. Other nodes can then try to contact these nodes in order announce to other nodes that they are available "behind" these nodes.

Since these nodes have limited resources, they should only allow a limited number number of relay requests.

## Desired feature

Define an upper bound of simultaneous relay connection and deny all further relay requests

- Add an additional error, showing the relay "is full"
- Handle this error at the client and try a different relay

### PR DESCRIPTION
**Pull request by andreykiryushkin**
Closes #208

This PR addresses an issue when relay is unaware of the situation when connection between nodes has been upgaded to webrtc and therefore reached it's capacity. A new status message `UPGRADED` was introduced to allow the relay to gracefully release the slot.

### CODE REVIEW

File: src/webrtc/connection.ts [hunk 1]
```diff
@@ -413,6 +422,8 @@ class WebRTCConnection implements MultiaddrConnection {
         }
 
         if (done) {
+          this.verbose(`FLOW: `)
+          this.relayConn.sendUpgrade()
```
Comment (robertkiel) [inline src/webrtc/connection.ts:426]: not sure whether this should come *before* or *after* the handover

**Comment by robertkiel** [comment]
Great PR! Addresses certainly one of the weak points

File: src/webrtc/connection.spec.ts [hunk 2]
```diff
@@ -52,6 +57,46 @@ describe('test webrtc connection', function () {
...
+
+      assert(u8aEquals((await BobShaker.read()).slice(), Uint8Array.from([MigrationStatus.NOT_DONE, ...firstMessage])))
+
+      const secondMessage = new TextEncoder().encode(`second message`)
+      BobShaker.write(Uint8Array.from([MigrationStatus.NOT_DONE, ...secondMessage]))
+
+      assert(u8aEquals((await AliceShaker.read()).slice(), secondMessage))
+    }
+
+    expect(sendUpgradedSpy).to.have.been.called.once
```
Comment (robertkiel) [inline src/webrtc/connection.spec.ts:97]: You probably not to `await` this

Comment (robertkiel) [reply src/webrtc/connection.spec.ts:97]: Used to use `p-defer` for this

File: src/relay/handshake.ts [hunk 3]
```diff
@@ -269,7 +270,7 @@ class RelayHandshake {
         this.shaker.rest()
         destinationShaker.rest()
 
-        createNew(source, destination, this.shaker.stream, destinationShaker.stream)
+        createNew(source, destination, this.shaker.stream, destinationShaker.stream, relayFreeTimeout)
```
Comment (robertkiel) [inline src/relay/handshake.ts:273]: why don't prefix with `__` to make sure that it is considered a variable that is only used for testing

File: src/relay/state.ts [hunk 4]
```diff
@@ -99,9 +99,9 @@ class RelayState {
    * @param toSource duplex stream to source
    * @param toDestination duplex stream to destination
    */
-  createNew(source: PeerId, destination: PeerId, toSource: Stream, toDestination: Stream) {
-    const toSourceContext = new RelayContext(toSource)
-    const toDestinationContext = new RelayContext(toDestination)
+  createNew(source: PeerId, destination: PeerId, toSource: Stream, toDestination: Stream, relayFreeTimeout?: number) {
+    const toSourceContext = new RelayContext(toSource, relayFreeTimeout)
```
Comment (robertkiel) [inline src/relay/state.ts:103]: I'd prefix `relayFreeTimeout` with `__`

File: scripts/integration-test.sh [hunk 5]
```diff
@@ -166,7 +166,7 @@ start_node tests/node.ts \
     --bootstrapPort ${charly_port} \
     --bootstrapIdentityName 'charly' \
     --noDirectConnections true \
-    --noWebRTCUpgrade true \
+    --noWebRTCUpgrade false \
```
Comment (robertkiel) [inline scripts/integration-test.sh:169]: I think we can also leave this option unconfigured, reduces the loc

Comment (andreykiryushkin) [reply scripts/integration-test.sh:169]: I prefer to be explicit in the tests so we don't hit some undefined behavior later when modifying tests

File: src/relay/context.ts [hunk 6]
```diff
@@ -267,6 +267,15 @@ class RelayContext extends EventEmitter {
 
             // close stream
             break
+          } else if (SUFFIX[0] == ConnectionStatusMessages.UPGRADED) {
+            // this is an artificial timeout to test the relay slot being properly freed during the integration test
+            this.verbose(`FLOW: waiting ${this.relayFreeTimeout}ms before freeing relay`)
+            await new Promise((resolve) => setTimeout(resolve, this.relayFreeTimeout))
```
Comment (robertkiel) [inline src/relay/context.ts:273]:
```suggestion
            if (this.relayFreeTimeout > 0) {
              await new Promise((resolve) => setTimeout(resolve, this.relayFreeTimeout))
            }
```

Comment (robertkiel) [reply src/relay/context.ts:273]: This should make sure that we don't rely on any Node.js event loop magic

File: src/webrtc/connection.ts [hunk 7]
```diff
@@ -298,10 +303,13 @@ class WebRTCConnection implements MultiaddrConnection {
         )
       )
     )
+    this.verbose(`FLOW: sending UPGRADED to relay`)
+    this.relayConn.sendUpgraded()
```
Comment (robertkiel) [inline src/webrtc/connection.ts:307]: What happens if WebRTC is unavailabe?

Comment (andreykiryushkin) [reply src/webrtc/connection.ts]: Right, I forgot this small fix. Pushed now. Thx.

**Review by robertkiel — APPROVED** [review]
lgtm, let's do some e2e reconnect test with some real `hoprd` nodes to make sure we don't break anything

**Pull request merged**